File: bindings/python/src/Experiment.cpp
#include <pybind11/pybind11.h>

#include <fort/myrmidon/Experiment.hpp>

namespace py = pybind11;

void BindExperiment(py::module_ & m) {
	using namespace fort::myrmidon;

	py::enum_<fort::tags::Family>(m,"TagFamily")
		.value("Tag36h11",fort::tags::Family::Tag36h11,"36h11 family")
		.value("Tag36h10",fort::tags::Family::Tag36h10,"36h10 family")
		.value("Tag36ARTag",fort::tags::Family::Tag36ARTag,"ARTag family")
		.value("Tag16h5",fort::tags::Family::Tag16h5,"16h5 family")
		.value("Tag25h9",fort::tags::Family::Tag25h9,"25h9 family")
		.value("Circle21h7",fort::tags::Family::Circle21h7,"Circle21h7 family")
		.value("Circle49h12",fort::tags::Family::Circle49h12,"Circle49h12 family")
		.value("Custom48h12",fort::tags::Family::Custom48h12,"Custom48h12 family")
		.value("Standard41h12",fort::tags::Family::Standard41h12,"Standard41h12 family")
		.value("Standard52h13",fort::tags::Family::Standard52h13,"Standard52h13 family")
		.value("Undefined",fort::tags::Family::Undefined,"Undefined family value");

	py::class_<Experiment>(m,
	                       "Experiment",
	                       "An Experiment is the main entry point of `fort-myrmidon`")
		.def(py::init(&fort::myrmidon::Experiment::Create),
		     py::arg("filepath"),
		     R"pydoc(
    Initialize an experiment from a specified filesystem
    location. This location will be used to determine relative path to
    the tracking data.

    Args:
        filepath (str): the wanted filesystem path to the experiment.
    Returns:
        py_fort_myrmidon.Experiment: a new empty Experiment associated with filepath
)pydoc")
		.def_static("Open",
		            &Experiment::Open,
		            py::arg("filepath"),
		            R"pydoc(
    Opens an existing Experiment on the filesystem

    Args:
        filepath (str): the filepath to open.
    Returns:
        py_fort_myrmidon.Experiment: the Experiment located at filepath
    Raises:
        Error: if filepath does not contains a valid Experiment or
            associated TrackingData is not accessible.
)pydoc")
		.def_static("OpenDataLess",
		            &Experiment::OpenDataLess,
		            py::arg("filepath"),
		            R"pydoc(
    Opens an existing Experiment on the filesystem in dataless mode.

    In dataless mode, no associated tracking data will be opened, but
    a TrackingSolver can be used to identify Ants in live tracking
    situation.

    Args:
        filepath (str): the filepath to open.
    Returns:
        py_fort_myrmidon.Experiment: the Experiment located at filepath
    Raises:
        Error: if filepath does not contains a valid Experiment
)pydoc")

		.def("Save",
		     &Experiment::Save,
		     py::arg("filepath"),
		     R"pydoc(
    Saves the experiment on the filesystem.

    Args:
        filepath: the filepath to save the experiment to.
    Raises:
        ValueError: if filepath would change the directory of the
            Experiment on the filesystem.
)pydoc")
		.def_property_readonly("AbsoluteFilePath",
		                       &Experiment::AbsoluteFilePath,
		                       " (str): the absolute filepath of the Experiment")
		.def("CreateSpace",
		     &Experiment::CreateSpace,
		     py::arg("name"),
		     py::return_value_policy::reference_internal,
		     R"pydoc(
    Creates a new Space in this Experiment.

    Args:
        name (str): the name for the new space
    Returns:
        py_fort_myrmidon.Space: the newly created Space
)pydoc")
		.def("DeleteSpace",
		     &Experiment::DeleteSpace,
		     py::arg("spaceID"),
		     R"pydoc(
    Deletes a Space from this Experiment.

    Args:
        spaceID (str): the spaceID of this space
    Raises:
        IndexError: if spaceID is not a valid SpaceID for this Experiment.
)pydoc")
		.def_property_readonly("Spaces",
		                       &Experiment::Spaces,
		                       py::return_value_policy::reference_internal,
		                       " (Dict[int,py_fort_myrmidon.Space]): this Experiment space indexed by their SpaceID")

		.def("AddTrackingDataDirectory",
		     &Experiment::AddTrackingDataDirectory,
		     py::arg("spaceID"),
		     py::arg("filepath"),
		     R"pydoc(
    Adds a tracking data directory to the Experiment.

    Args:
        spaceID (int): the space to add the tracking data directory
            to.
        filepath (str): the filepath to the tracking data directory.
    Returns:
        str: the relative path from self.AbsoluteFilePath to filepath,
            that will be the URI to identify the tracking data
            directory.
    Raises:
        IndexError: if spaceID is not valid for this Experiment
        RuntimeError: if filepath is not a valid tracking data directory.
        RuntimeError: if the data will overlap in time with another directory in the same space
        RuntimeError: if the data is used by another space
        ValueError: if the tag family does not match with other directory in the experiment
)pydoc")
		.def("RemoveTrackingDataDirectory",
		     &Experiment::RemoveTrackingDataDirectory,
		     py::arg("URI"),
		     R"pydoc(
    Removes a tracking data directory from the Experiment.

    Args:
        URI (str): the URI that identifies the tracking data directory
    Raises:
        IndexError: if URI does not identifies a tracking data
            directory in this experiment.
)pydoc")
		.def("CreateAnt",
		     &Experiment::CreateAnt,
		     R"pydoc(
    Creates a new Ant in the Experiment.

    Returns:
        py_fort_myrmidon.Ant: the newly created Ant
)pydoc")
		.def("DeleteAnt",
		     &Experiment::DeleteAnt,
		     py::arg("antID"),
		     R"pydoc(
    Deletes an Ant from the Experiment

    Args:
        antID (int): the AntID of the Ant to remove
    Raises:
        IndexError: if antID is invalid for the experiment
        RuntimeError: if the ant still have Identification targetting
                      her
)pydoc")
		.def("AddIdentification",
		     &Experiment::AddIdentification,
		     py::arg("antID"),
		     py::arg("tagID"),
		     py::arg("start") = fort::Time::SinceEver(),
		     py::arg("end") = fort::Time::Forever(),
		     R"pydoc(
    Adds an Identification to the Experiment

    Args:
        antID (int): the ant to target
        tagID (int): the TagID to use
        start (py_fort_myrmidon.Time): the first valid Time for the
        Identification
        end (py_fort_myrmidon.Time): the first invalid Time for the
        Identification

    Returns:
        py_fort_myrmidon.Identification: the newly created
            Identification

    Raises:
        IndexError: if antID is not valid for the Experiment

        RuntimeError: if this identification would overlap in time
                      with another one, either for a given antID or
                      tagID.
)pydoc")
		.def("DeleteIdentification",
		     &Experiment::DeleteIdentification,
		     py::arg("identification"),
		     R"pydoc(
    Deletes an Identification from this Experiment

    Args:
        identification (py_fort_myrmidon.Identification):
            the Identification to remove

    Raises:

        ValueError: if identification is not a valid Identification
                    object from this Experiment.
)pydoc")
		.def("FreeIdentificationRangeAt",
		     &Experiment::FreeIdentificationRangeAt,
		     py::arg("tagID"),
		     py::arg("time"),
		     R"pydoc(
    Computes a valid Identification time range for a tag

    Args:
        tagID (int): the tagID to query for
        time (py_fort_myrmidon.Time): a time that must be contained in
            the result.

    Returns:
        py_fort_myrmidon.Time: the first valid time, can be
            Time.SinceEver()
        py_fort_myrmidon.Time: the first invalid time, can be
            Time.Forever()

   Raises:
        RuntimeError: if tagID already identifies an Ant at time
)pydoc")
		.def_property("Name",
		              &Experiment::Name,
		              &Experiment::SetName,
		              " (str): the name of the Experiment")
		.def_property("Author",
		              &Experiment::Author,
		              &Experiment::SetAuthor,
		              " (str): the author of the Experiment")
		.def_property("Comment",
		              &Experiment::Comment,
		              &Experiment::SetComment,
		              " (str): a comment about the Experiment")
		.def_property_readonly("Family",
		                       &Experiment::Family,
		                       " (py_fort_myrmidon.TagFamily): the TagFamily used in the Experiment")
		.def_property("DefaultTagSize",
		              &Experiment::DefaultTagSize,
		              &Experiment::SetDefaultTagSize,
		              " (float): the default tag size in mm used in the Experiment")

		;
}

File: bindings/R/FortMyrmidon/R/experiment.R
#' @name fmExperiment
#' @title API entry point
#' @description An fmExperiment olds a collection of
#'     \code{\link{fmAnt}}, \code{\link{fmIdentification}},
#'     \code{\link{fmSpace}} and \code{\link{fmZone}} and give access
#'     to the identified tracking data instantaneous collision and
#'     interaction detection through \code{\link{fmQuery}}.
#' @description fmExperiment are saved to the filesystem in
#'     `.myrmidon` files. One can use \code{\link{fmExperimentOpen}()}
#'     and \code{\link{fmExperiment$save}()} to interact with these
#'     files.
#' @description One can use
#'     \code{\link{fmExperiment$addTrackingDataDirectory}()} to link
#'     an fmExperiment with some tracking data, organized by
#'     **fort-leto** in a tracking data directory. This data must be
#'     assigned to a \code{\link{fmSpace}} (previously created with
#'     \code{\link{fmExperiment$createSpace}()}).  fmExperiment saves
#'     relative links to these tracking data directory. These paths
#'     are relative, so one can rename a `.myrmidon` file on the
#'     filesystem with \code{\link{fmExperiment$save}()}, but it must
#'     remains in the same directory.
#'@description In **fort-myrmidon**, tags are not used
#'     directly. Instead user are required to make \code{\link{fmAnt}}
#'     object (through \code{\link{fmExperiment$createAnt}()}) and use
#'     \code{\link{fmIdentification}} (through
#'     \code{\link{fmExperiment$addIdentification}()}) to relate a tag
#'     value to an \code{\link{fmAnt}}. To perform collision and
#'     interaction detection, users can create for each Ant a virtual
#'     shape, made of a collection of \code{\link{fmCapsule}}. Each
#'     Capsule is assigned an AntShapeTypeID (an integer starting from
#'     1) which must be previously defined using
#'     \code{\link{fmExperiment$createAntShapeType}()}. There is no
#'     predefined AntShapeTypeID.
#' @description **fort-studio** allows to make measurement on close-up
#'     of each \code{\link{fmAnt}}. These measurement must be assigned
#'     to a type, created with
#'     \code{\link{fmExperiment$createMeasurementType}()}. There is a
#'     predefined, undeletable MeasurementTypeID:
#'     \code{\link{fmExperiment.HEAD_TAIL_MEASUREMENT_TYPE_ID}}. It is
#'     used to automatically determine \code{$antPosition} and
#'     \code{$antAngle} fields of \code{\link{fmIdentification}} from
#'     **fort-studio** measurements.
#' @description Each \code{\link{fmAnt}} can also holds a dictionnary
#'     of key/value pairs. The key name, type and initial value for
#'     each Ant must be defined with
#'     \code{\link{fmExperiment$setMetaDataKey}()}. Through
#'     \code{\link{fmAnt$setValue}()}, individual, timed value can be
#'     assigned to each Ant. There are no predefined keys.
#' @field name a character that holds the fmExperiment name.
#' @field author a character that holds the fmExperiment author.
#' @field comment a character that holds the fmExperiment comment.
#' @field spaces a read-only, list of fmSpace of this fmExperiment.
#' @field ants a read-only, list of fmAnt of this fmExperiment.
#' @field absoluteFilePath a read-only, character holding the
#'     associated filepath of this fmExperiment.
#' @field defaultTagSize the default tag size in mm for
#'     fmIdentification in this fmExperiment.
#' @field mesurementTypeNames a read-only list of the fmExperiment
#'     defined measurement types
#' @field antShapeTypeNames a read-only list of the fmExperiment
#'     defined ant shape types
#' @field metaDataKeys a read-only list of the fmExperiment user
#'     defined metadata keys.
#' @family fmExperiment methods
NULL

#' @name fmExperiment$save
#' @title Saves the fmExperiment to the filesystem
#' @description Saves the fmExperiment to the filesystem. Errors
#'     occurs if this method is used to change the directory of the
#'     current fmExperiment.
#' @param filepath the filepath to save the fmExperiment to
#' @family fmExperiment methods
NULL

#' @name fmExperiment$createSpace
#' @title Creates a new fmSpace for this fmExperiment
#' @description Create a new \code{\link{fmSpace}} for this fmExperiment
#' @param name the name for the new fmSpace
#' @return the newly created \code{\link{fmSpace}}
#' @family fmExperiment methods
NULL

#' @name fmExperiment$deleteSpace
#' @title Deletes a fmSpace from this fmExperiment
#' @description Deletes a \code{\link{fmSpace}} from this
#'     fmExperiment. An error occurs if **spaceID** is not valid for
#'     this fmExperiment.
#' @param spaceID the $ID of the fmSpace to remove
#' @family fmExperiment methods
NULL

#' @name fmExperiment$addTrackingDataDirectory
#' @title Associates a tracking data directory with this fmExperiment.
#' @description Associates a tracking data directory to a
#'     \code{\link{fmSpace}} of this fmExperiment. An error could
#'     occur if either: \itemize{ \item **spaceID** is not valid for
#'     this fmExperiment.  \item **filepath** is not a valid tracking
#'     data directory \item **filepath** data would overlap in Time
#'     with data from another directory in **spaceID** \item
#'     **filepath** is associated with another fmSpace \item
#'     **filepath** is using a tag family that is different from the
#'     other directory associated in this experiment.  }
#' @param spaceID the $ID of the fmSpace to associate data with
#' @param filepath the filepath to the tracking data directory
#' @return the URI used to designate this directory in this fmExperiment
#' @family fmExperiment methods
NULL

#' @name fmExperiment$removeTrackingDataDirectory
#' @title Removes a tracking data directory from this fmExperiment.
#' @description Removes a tracking data directory from this
#'     fmExperiment. An error occurs if **URI** does not designate a
#'     tracking data directory from this fmExperiment.
#' @param URI the URI of the tracking data directory to remove
#' @family fmExperiment methods
NULL


#' @name fmExperiment$createAnt
#' @title Creates a new fmAnt for this fmExperiment
#' @description Creates a new \code{\link{fmAnt}} for this
#'     fmExperiment
#' @return the newly create \code{\link{fmAnt}}
#' @family fmExperiment methods
NULL


#' @name fmExperiment$deleteAnt
#' @title Removes a fmAnt from this fmExperiment
#' @description Removes a \code{\link{fmAnt}} from this
#'     fmExperiment. An error occurs if **antID** is not valid for
#'     this fmExperiment.
#' @param antID the $ID of the \code{\link{fmAnt}} to remove.
#' @family fmExperiment methods
NULL

#' @name fmExperiment$addIdentification
#' @title Adds an fmIdentication to this fmExperiment
#' @description Adds a \code{\link{fmIdentification}} to this
#'     fmExperiment. An error occurs if any time overlap with other
#'     fmIdentification would occurs for **antID** or for **tagID**.
#' @param antID the $ID of the \code{\link{fmAnt}} to identify.
#' @param tagID the tag value to associate with **antID**.
#' @param start the first valid time for this fmIdentification. Could
#'     be \code{\link{fmTimeSinceEver}()}.
#' @param end the first invalid time for this fmIdentification. Could
#'     be \code{\link{fmTimeForever}()}.
#' @return the newly created \code{\link{fmIdentification}}
#' @family fmExperiment methods
NULL

#' @name fmExperiment$deleteIdentification
#' @title Removes an fmIdentication from this fmExperiment
#' @description Removes a \code{\link{fmIdentification}} from this
#'     fmExperiment. An error occurs if the passed object is not valid
#'     for this fmExperiment.
#' @param identification the fmIdentification to remove
#' @family fmExperiment methods
NULL

#' @name fmExperiment$freeIdentificationRangeAt
#' @title Returns an available time range for a tag.
#' @description Returns an available time range for a tag. An error
#'     occurs if **tagID** is used in an fmIdentification at **time**.
#' @param tagID the tag value to query for
#' @param time the \code{\link{fmTime}} to find a range for.
#' @return a named list ($start,$end) containing the available time
#'     range containing **time**.
#' @family fmExperiment methods
NULL

#' The default available measurement type of any fmExperiment
#' @family fmExperiment methods
fmExperiment.HEAD_TAIL_MEASUREMENT_TYPE_ID <- 1L

#' @name fmExperiment$createMeasurementType
#' @title Create a new measurement type for this fmExperiment.
#' @description Create a new measurement type for this fmExperiment.
#' @param name the name for the new measurement type
#' @return the new measurement type ID that will identifies this
#'     measurement type.
#' @family fmExperiment methods
NULL

#' @name fmExperiment$setMeasurementTypeName
#' @title Renames a measurement type for this fmExperiment.
#' @description Renames a measurement type for this fmExperiment. An
#'     error occurs if **typeID** is not valid for this fmExperiment.
#' @param typeID the measurement typeID to rename.
#' @param name the new name for the measurement type.
#' @family fmExperiment methods
NULL

#' @name fmExperiment$deleteMeasurementType
#' @title Removes a measurement type for this fmExperiment.
#' @description Removes a measurement type for this fmExperiment. An
#'     error occurs if **typeID** is not valid for this fmExperiment
#'     or if the fmExperiment contains measurement for this type or if
#'     \code{\link{fmExperiment.HEAD_TAIL_MEASUREMENT_TYPE_ID}} is
#'     removed.
#' @param typeID the measurement typeID to remove.
#' @family fmExperiment methods
NULL


#' @name fmExperiment$createAntShapeType
#' @title Create a new ant shape type for this fmExperiment.
#' @description Create a new ant shape type for this fmExperiment.
#' @param name the name for the new ant shape type
#' @return the new ant shape type ID that will identifies this
#'     ant shape type.
#' @family fmExperiment methods
NULL

#' @name fmExperiment$setAntShapeTypeName
#' @title Renames an ant shape type for this fmExperiment.
#' @description Renames an ant shape type for this fmExperiment. An
#'     error occurs if **shapeTypeID** is not valid for this fmExperiment.
#' @param shapeTypeID the ant shape typeID to rename.
#' @param name the new name for the ant shape type.
#' @family fmExperiment methods
NULL

#' @name fmExperiment$deleteAntShapeType
#' @title Removes an ant shape type for this fmExperiment.
#' @description Removes a measurement type for this fmExperiment. An
#'     error occurs if **shapeTypeID** is not valid for this fmExperiment
#'     or if at least one fmAnt contains a fmCapsule of this type.
#' @param typeID the ant shape type ID to remove.
#' @family fmExperiment methods
NULL

#' @name fmExperiment$setMetaDataKey
#' @title Sets or creates the name, type and default value for an user
#'     defined metadata key.
#' @description Sets or creates the name, type and default value for
#'     an user defined metadata key. the type is inferred from
#'     **defaultValue** which should either be a logical, integer,
#'     numerical, character or \code{\link{fmTime}} value. An error
#'     could occurs if **key** is already defined, **defaultValue**
#'     would change the type for **key** and at least one fmAnt has a
#'     defined timed data for **key**.
#' @param key the key to define
#' @param defaultValue the default value for **key**
#' @family fmExperiment methods
NULL

#' @name fmExperiment$renameMetaDataKey
#' @title Renames a metadata key.
#' @description Renames a metadata key. An error occurs if **oldKey**
#'     is unknown or if **newKey** is already used.
#' @param oldKey the key to rename
#' @param newKey the new name for the key
#' @family fmExperiment methods
NULL

#' @name fmExperiment$deleteMetaDataKey
#' @title Deletes a metadata key.
#' @description Deletes a metadata key. An error occurs if **key** is
#'     unknown or if at least one Ant contains timed data for **key**.
#' @param key the key to remove
#' @family fmExperiment methods
NULL

#' @name fmExperiment$identificationsAt
#' @title Obtains the AntID <-> TagID correspondances for a time
#' @description Obtains the AntID <-> TagID correspondances for a
#'     given **time** as a list. Index of the list are the AntID, and
#'     the value the corresponding TagID. If an AntID does not exist
#'     or is not identified by any TagID, the corresponding value in
#'     the list will be NULL.
#' @return a list of TagID indexed by AntID.
#' @param time the time to query identification for.
#' @family fmExperiment methods
NULL

#' Enumerates the possible tag family used in the FORT system.
fmTagFamily <- c("Tag36h11" = 0,
                 "Tag36h10" = 1,
                 "Tag36ARTag" = 2,
                 "Tag16h5" = 3,
                 "Tag25h9" = 4,
                 "Circle21h7" = 5,
                 "Circle49h12" = 6,
                 "Custom48h12" = 7,
                 "Standard41h12" = 8,
                 "Standard52h13" = 9,
                 "Undefined" = 10)
